### PR TITLE
[heft] Feature: RegisterParameters on any Action

### PR DESCRIPTION
**Pull request by halfnibble**
## Summary

The goal is to make it trivial for Heft plugins to add custom parameters to built-in or other existing Heft actions. 

An example plugin can be found in `build-tests/heft-parameter-plugin`

## Details

Here is some example usage: 

```typescript
class HeftParameterPlugin implements IHeftPlugin {
  public readonly pluginName: string = 'heft-action-plugin';

  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
    const customParameter: IHeftFlagParameter = heftSession.commandLine.registerFlagParameter({
      associatedActionNames: ['build', 'test', 'start'],
      parameterLongName: '--custom-parameter',
      description: 'Test running a custom parameter'
    });

    const customStringParameter: IHeftStringParameter = heftSession.commandLine.registerStringParameter({
      associatedActionNames: ['build', 'test', 'start'],
      parameterLongName: '--custom-string-parameter',
      description: 'Test running a custom string parameter',
      argumentName: 'TEXT',
      required: true
    });

    heftSession.hooks.build.tap(this.pluginName, (build: IBuildStageContext) => {
      build.hooks.compile.tap(this.pluginName, (compile: ICompileSubstage) => {
        compile.hooks.run.tapPromise(this.pluginName, async () => {
          if (customParameter.actionAssociated && customStringParameter.valueProvided) {
            // Do something with `customStringParameter.value`
          }
        });
      });
    });
    ...
````

## How it was tested

There is a test project here: `build-tests/heft-parameter-plugin-test`

### CODE REVIEW

File: apps/heft/src/cli/HeftToolsCommandLineParser.ts
Comment (octogonz) [inline apps/heft/src/cli/HeftToolsCommandLineParser.ts:125]: What happens if two different plugins try to register the same parameter name (or action)?

🤔 We could imagine a situation where two plugins were developed independently and never tested together. Renaming a parameter is a breaking change, so it might be difficult to get the plugin author to release a fix. We could imagine a Heft feature that remaps the names automatically (e.g. `--my-option` --> `--my-plugin-name-my-option`) -- or maybe it's safer to require the user to remap the name manually using a config setting.

But let's say that we won't implement that initially.  Instead, Heft will simply abort and print an error.  What does the error message look like?  Does it at least indicate:

- what is the parameter name causing the conflict?
- what are the names of the plugins that are trying to define this parameter?

Comment (octogonz) [reply apps/heft/src/cli/HeftToolsCommandLineParser.ts:125]: (The same considerations apply to `registerAction()`)

File: apps/heft/src/cli/actions/CustomAction.ts
Comment (halfnibble) [reply apps/heft/src/cli/actions/CustomAction.ts:143]: This was removed for the same reason as above (the value in default is `never` because all valid cases are handled). So whatever we decide on the above case, I'll apply here too.

File: apps/heft/src/cli/HeftToolsCommandLineParser.ts
Comment (D4N14L) [reply apps/heft/src/cli/HeftToolsCommandLineParser.ts:175]: It would be good if we were retroactively prepared for that failure case.

File: apps/heft/src/cli/HeftToolsCommandLineParser.ts
Comment (halfnibble) [reply apps/heft/src/cli/HeftToolsCommandLineParser.ts:125]: For registerParameters, there is a check on line 131 that prints:
`throw new Error(`Duplicate parameter name: ${parameterName}`);`

But maybe it should print the action name too. Also, we could add a registry of the plugins that introduce the parameters, and then we could include the name of the conflicting plugins in the error message. 

Remapping parameters seems confusing. Ultimately, we should discourage use of parameter and action names that are likely to conflict. 

I think registerAction currently just overrides existing actions, but I have not tested it yet.

File: apps/heft/src/cli/actions/CustomParameters.ts
Comment (iclanton) [inline apps/heft/src/cli/actions/CustomParameters.ts:51]: Instead of a callback, it feels like it would be more natural to have return an object with typed parameter values that throws an error upon access if the parameters haven't been parsed yet.

File: apps/heft/src/cli/HeftToolsCommandLineParser.ts
Comment (iclanton) [inline apps/heft/src/cli/HeftToolsCommandLineParser.ts:123]: This should throw if it's called after the parameters have been parsed.

File: apps/heft/src/cli/actions/HeftActionBase.ts
Comment (halfnibble) [reply apps/heft/src/cli/actions/HeftActionBase.ts:188]: Had to convert the `Set()` to and `Array`, but this should work, right?
```typescript
await Promise.all(
  [...this.customParametersCallbacks].map(async (callback) => await callback(customParametersRecord))
);
```

File: build-tests/heft-parameter-plugin-test/package.json
Comment (halfnibble) [reply build-tests/heft-parameter-plugin-test/package.json:8]: Unfortunately, not at this time, as we discussed. :(
Maybe a future version of ts-command-line will support it.

File: common/reviews/api/heft.api.md
Comment (halfnibble) [reply common/reviews/api/heft.api.md:61]: Undoing this.

File: apps/heft/src/cli/actions/HeftActionBase.ts
Comment (D4N14L) [reply apps/heft/src/cli/actions/HeftActionBase.ts:188]: Array map is fine, but just make it:
```js
await Promise.all(
  [...this.customParametersCallbacks].map((callback) => callback(customParametersRecord))
);
```

File: build-tests-samples/heft-storybook-react-tutorial-storykit/package.json
Comment (halfnibble) [reply build-tests-samples/heft-storybook-react-tutorial-storykit/package.json]: Fixed

File: apps/heft/src/cli/actions/HeftActionBase.ts
Comment (D4N14L) [inline apps/heft/src/cli/actions/HeftActionBase.ts:190]: See unresolved comment, this won't work quite right either.

File: apps/heft/src/cli/actions/HeftActionBase.ts
Comment (halfnibble) [reply apps/heft/src/cli/actions/HeftActionBase.ts]: Thanks for the correction. I'm going to mark this resolved as this bit of code is now removed.

File: apps/heft/src/cli/actions/CustomParameters.ts
Comment (halfnibble) [reply apps/heft/src/cli/actions/CustomParameters.ts]: Done. Thanks for the suggestion.

File: apps/heft/src/cli/HeftToolsCommandLineParser.ts
Comment (halfnibble) [reply apps/heft/src/cli/HeftToolsCommandLineParser.ts:123]: Fixed.

File: apps/heft/src/cli/actions/HeftActionBase.ts
Comment (halfnibble) [reply apps/heft/src/cli/actions/HeftActionBase.ts]: Thanks. This callback code is now replaced with something a bit better.

File: apps/heft/src/cli/HeftToolsCommandLineParser.ts
Comment (D4N14L) [inline apps/heft/src/cli/HeftToolsCommandLineParser.ts:226]: It's not really a record at this point, especially since you're defining getter properties on it below. Maybe `parametersObject`

File: apps/heft/src/cli/HeftToolsCommandLineParser.ts
Comment (iclanton) [inline apps/heft/src/cli/HeftToolsCommandLineParser.ts:130]:
```suggestion
              `Unable to register parameters for action "${action.actionName}" after parameters have already been processed`
```

File: libraries/ts-command-line/src/providers/CommandLineParameterProvider.ts
Comment (D4N14L) [inline libraries/ts-command-line/src/providers/CommandLineParameterProvider.ts:353]: Might be smart to set this after processing instead of before.

**Review by D4N14L — APPROVED** [review]
Signing off, but I'd wait for @iclanton review before merge :)

File: apps/heft/src/cli/test/__snapshots__/CommandLineHelp.test.ts.snap
Comment (D4N14L) [inline apps/heft/src/cli/test/__snapshots__/CommandLineHelp.test.ts.snap:70]: Will this info show up when using the storybook plugin? I think it should given how things are formatted, but it's not quite clear.

**Review by iclanton — APPROVED** [review]

File: apps/heft/src/cli/test/__snapshots__/CommandLineHelp.test.ts.snap
Comment (halfnibble) [reply apps/heft/src/cli/test/__snapshots__/CommandLineHelp.test.ts.snap:70]: I verified that this shows up in the storybook-react-tutorial sample when running `heft start -h`

**Pull request merged**
